README.md
# cryptor

AES-GCM-256 algorithm cryptor

## Usage

```dart
  var text = 'SOME DATA TO ENCRYPT';
  var password = 'password';

  var encrypted = Cryptor.encrypt(text, password);
  var decrypted = Cryptor.decrypt(encrypted, password);

  print(encrypted);
  print(decrypted);
```


CHANGELOG.md
## [0.1.6] - Upgrade steel_crypt version

- Upgrade steel_crypt version to 2.0.0

## [0.1.5] - Fix pbkdf2 and improvements

- Fix pbkdf2 to generate correct key
- Reduce number of type conversions
- Add hash function to hash password

## [0.1.4] - API Change

- encrypt and decrypt are now static methods of Cryptor
- Implement AesGcm module

## [0.1.3] - Implement PBKDF2

- Implement PBKDF2
- Get rid of hash_password dependency

## [0.1.2] - Initial implementation of AES GCM cryptor

- Rename example file to be inline with guidelines

## [0.1.1] - Initial implementation of AES GCM cryptor

- Add dart doc comments to the code
- Add expample/aes.dart
- Formatted lib/cryptor.dart

## [0.1.0] - Initial implementation of AES GCM cryptor

- Add `ecnrypt` and `decrypt` functions


test/pbkdf2_test.dart
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';

import 'package:crypto/crypto.dart';
import 'package:cryptor/pbkdf2.dart';

void main() {
  List<int> key;
  final PBKDF2 generator = new PBKDF2(sha512);
  final String password = 'password';
  final Uint8List salt = Uint8List.fromList([
    49,
    3,
    239,
    154,
    144,
    119,
    1,
    179,
    39,
    95,
    202,
    162,
    85,
    242,
    222,
    240,
    197,
    203,
    194,
    141,
    242,
    46,
    34,
    180,
    165,
    241,
    242,
    163,
    227,
    141,
    183,
    247,
    76,
    85,
    18,
    139,
    255,
    190,
    142,
    241,
    178,
    4,
    183,
    204,
    58,
    7,
    161,
    84,
    251,
    181,
    210,
    218,
    36,
    65,
    207,
    112,
    38,
    251,
    11,
    72,
    196,
    30,
    3,
    143
  ]);

  test('Derives 16 bit key from password and salt', () {
    key = generator.generateKey(password, salt, 100000, 16);
    expect(key, [
      38,
      11,
      27,
      194,
      233,
      120,
      192,
      88,
      137,
      233,
      67,
      73,
      165,
      92,
      41,
      112
    ]);
  });

  test('Derives 32 bit key from password and salt', () {
    key = generator.generateKey(password, salt, 100000, 32);
    expect(key, [
      38,
      11,
      27,
      194,
      233,
      120,
      192,
      88,
      137,
      233,
      67,
      73,
      165,
      92,
      41,
      112,
      168,
      252,
      140,
      24,
      239,
      79,
      222,
      27,
      33,
      10,
      158,
      88,
      77,
      11,
      144,
      208
    ]);
  });

  test('Derives 64 bit key from password and salt', () {
    key = generator.generateKey(password, salt, 100000, 64);
    expect(key, [
      38,
      11,
      27,
      194,
      233,
      120,
      192,
      88,
      137,
      233,
      67,
      73,
      165,
      92,
      41,
      112,
      168,
      252,
      140,
      24,
      239,
      79,
      222,
      27,
      33,
      10,
      158,
      88,
      77,
      11,
      144,
      208,
      46,
      105,
      9,
      4,
      127,
      52,
      139,
      201,
      126,
      76,
      17,
      199,
      45,
      182,
      68,
      23,
      78,
      17,
      85,
      235,
      240,
      34,
      167,
      254,
      248,
      101,
      106,
      23,
      144,
      249,
      203,
      175
    ]);
  });
}


test/cryptor_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:cryptor/cryptor.dart';

void main() {
  test('Encrypts and decrypts string', () {
    final password = 'password';
    String encrypted = Cryptor.encrypt('data', password);
    expect(Cryptor.decrypt(encrypted, password), 'data');
  });

  test('Decrypts data encrypted on other platform', () {
    final encrypted =
        'NhhXJJ+bhoQCSTcTWNW25AtmmVGGfBXEkYdldy/TK3zluwu40LIPmWbpXtw1ntnwx/ZfNJPeGlFlh1s+QawX8NgTNJfBVfW1s8l4DO9w+1kWiYvlLX3klaNSfPEuOHvKMpnyJ+t6p+j3NiBHcAvbKmYCvLkUHGgSh9r0nCQuADnvyvjo3RO0BjegxM/GYze8nQupQQYeeshrF19bOQ4Wmy9mDScWuqY9cRMNDd+BRuWWFi1tpXUGNc5yblHW4qABW4FR1r+Hgt4MwfaYfKO4gc8+tmPnf/zvtXZDeoKaXIOy1+Rn6uN8aNoZSEohn4fsyOpUDhO+KWpiJSCWxipnQeuqoeALG+af/5qgxa0HFcUeg7vxlGB8YgK/3nbproN/3/74EWANziHodS0A/E+OM9X+ltVAeaSk3DfMxzHH178lia6RDrxuh4DElgh6b55uejhKrxrCxTzSkn8TIQkspWse47VdCXSrlQukdh7n2qjKWmc06CbXqzsoN6BHJ4Aw/l0KyiFveimG1jB8PFPPgATf9gNKuQ3YinvYrv141zJB1kpGjeSu7mRc/oYyLofoeLhjwHPF/QR4G5EuhFUjWpPr4NuRJgGBLefdZJGjJkeOMjkFDCQ6GorFlwCc9Lc6OAUVJ/JDOXaoLqexnTrAFHmjxy05mgzY6JzHpP+4xmg+g9wjJXyFEx3EX9HHD2ddmYLjFLroA36SmGZoegzb/UhOKKpEHtC/Yu8EPiwgVhkXIVBRZvVRh5jBS7LQkyI6KmDs7vmvutUq3zt109rxWQ==';
    final password =
        'sQnzu7wkTrgkQZF+0G1hi5AI3Qmzvv0bXgc5THBqi7mAsdd4Xll27ASbRt9fEyavWi6m0QP9B8lThf+rDKy8hg==';
    expect(Cryptor.decrypt(encrypted, password),
        '{"entries":[{"id":"MSz3VJcdQ","type":"login","title":"Facebook","website":"https://facebook.com","username":"alchaplinsky","password":"rUHmTveu8a6IGg4wK2GYcq5jsowzSoSPwQpn9ibT1Kb/VaRbwa5c+UneHmPm0B/whZk0mfRsodxpPkXoiatlmSN7Q5K3b+Wrb1TXkQ5EKk905kcPsIa6DsJyNBGmmX7WavZC76h5W2vjk/SKhaQM/y/ValtecFU8Zs43P0PcM78=","email":"","note":"","otp":"","password_updated_at":"2020-05-23T23:59:24.592+02:00","createdAt":"2020-05-23T23:59:25.575+02:00","updatedAt":"2020-05-23T23:59:52.796+02:00"}]}');
  });

  test('Hashes password', () {
    expect(Cryptor.hash('password'),
        'sQnzu7wkTrgkQZF+0G1hi5AI3Qmzvv0bXgc5THBqi7mAsdd4Xll27ASbRt9fEyavWi6m0QP9B8lThf+rDKy8hg==');
  });
}


lib/pbkdf2.dart
import 'dart:convert';
import 'dart:typed_data';
import 'dart:math';

import 'package:crypto/crypto.dart';

/// Instance of PBKDF2 derive a key from password, salt, and hash function.
///
/// https://en.wikipedia.org/wiki/PBKDF2
class PBKDF2 {
  Hash _hashAlgorithm = sha256;
  int _blockSize;

  /// Creates instance capable of generating a key.
  ///
  /// [hashAlgorithm] defaults to [sha256].
  PBKDF2([Hash hashAlgorithm]) {
    _hashAlgorithm = hashAlgorithm;
    _blockSize = _hashAlgorithm.convert([1, 2, 3]).bytes.length;
  }

  /// Hashes a [password] with a given [salt].
  ///
  /// The length of this return value will be [keyLength].
  Uint8List generateKey(
      String password, Uint8List salt, int rounds, int keyLength) {
    _validateKeyLength(keyLength);

    int numberOfBlocks = (keyLength / _blockSize).ceil();
    Hmac hmac = new Hmac(_hashAlgorithm, utf8.encode(password));
    ByteData key = new ByteData(keyLength);
    int offset = 0;

    int saltLength = salt.length;
    ByteData inputBuffer = new ByteData(salt.length + 4)
      ..buffer.asUint8List().setRange(0, salt.length, salt);

    for (int blockNumber = 1; blockNumber <= numberOfBlocks; blockNumber++) {
      inputBuffer.setUint8(saltLength, blockNumber >> 24);
      inputBuffer.setUint8(saltLength + 1, blockNumber >> 16);
      inputBuffer.setUint8(saltLength + 2, blockNumber >> 8);
      inputBuffer.setUint8(saltLength + 3, blockNumber);

      Uint8List block = _XORDigestSink.generate(inputBuffer, hmac, rounds);
      int blockLength = _blockSize;
      if (offset + blockLength > keyLength) {
        blockLength = keyLength - offset;
      }
      key.buffer.asUint8List().setRange(offset, offset + blockLength, block);

      offset += blockLength;
    }
    return key.buffer.asUint8List();
  }

  void _validateKeyLength(int keyLength) {
    if (keyLength > (pow(2, 32) - 1) * _blockSize) {
      throw new Exception('PBKDF2: Derived key is too long');
    }
  }
}

class _XORDigestSink extends Sink<Digest> {
  _XORDigestSink(ByteData inputBuffer, Hmac hmac) {
    lastDigest = hmac.convert(inputBuffer.buffer.asUint8List()).bytes;
    bytes = new ByteData(lastDigest.length)
      ..buffer.asUint8List().setRange(0, lastDigest.length, lastDigest);
  }

  static Uint8List generate(ByteData inputBuffer, Hmac hmac, int rounds) {
    _XORDigestSink hashSink = new _XORDigestSink(inputBuffer, hmac);

    // If rounds == 1, we have already run the first hash in the constructor
    // so this loop won't run.
    for (int round = 1; round < rounds; round++) {
      ByteConversionSink hmacSink = hmac.startChunkedConversion(hashSink);
      hmacSink.add(hashSink.lastDigest);
      hmacSink.close();
    }

    return hashSink.bytes.buffer.asUint8List();
  }

  ByteData bytes;
  Uint8List lastDigest;

  @override
  void add(Digest digest) {
    lastDigest = digest.bytes;
    for (int i = 0; i < digest.bytes.length; i++) {
      bytes.setUint8(i, bytes.getUint8(i) ^ lastDigest[i]);
    }
  }

  @override
  void close() {}
}


lib/aes_gcm.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:steel_crypt/PointyCastleN/export.dart';

class AesGcm {
  Uint8List _key32;

  AesGcm(this._key32);

  String encrypt(String input, [Uint8List iv]) {
    CipherParameters params = PaddedBlockCipherParameters(
        ParametersWithIV<KeyParameter>(KeyParameter(_key32), iv), null);
    PaddedBlockCipher cipher = PaddedBlockCipher('AES/GCM/PKCS7');
    cipher..init(true, params);
    Uint8List inter = cipher.process(utf8.encode(input) as Uint8List);
    return base64.encode(inter);
  }

  String decrypt(String encrypted, [Uint8List iv]) {
    CipherParameters params = PaddedBlockCipherParameters(
        ParametersWithIV(KeyParameter(_key32), iv), null);
    PaddedBlockCipher cipher = PaddedBlockCipher('AES/GCM/PKCS7');
    cipher..init(false, params);
    Uint8List inter = cipher.process(base64.decode(encrypted));
    return utf8.decode(inter);
  }
}


lib/cryptor.dart
library cryptor;

import 'dart:convert';
import 'dart:typed_data';
import 'dart:math';

import 'package:crypto/crypto.dart';

import './pbkdf2.dart';
import './aes_gcm.dart';

class Cryptor {
  static const SALT_LENGTH = 64;
  static const IV_LENGTH = 16;
  static const TAG_LENGTH = 16;
  static const KEY_LENGTH = 32;
  static const KEY_ITERATIONS_COUNT = 100000;

  /// Hashes passed [value] with sha512 algorithm
  static String hash(String value) {
    return base64.encode(sha512.convert(utf8.encode(value)).bytes);
  }

  /// Encrypts passed [cleartext] with key generated based on [password] argument
  static String encrypt(String cleartext, String password) {
    Uint8List salt = _generateSalt();
    Uint8List iv = _generateIV();
    Uint8List key = _pbkdf2(password, salt);

    String ciphertext = AesGcm(key).encrypt(cleartext, iv);
    return base64.encode(_combineInputs(salt, iv, _generateTag(), ciphertext));
  }

  /// Encrypts passed [cryptedtext] with key generated based on [password] argument
  static String decrypt(String cryptedtext, String password) {
    Uint8List bytes = base64.decode(cryptedtext);
    Uint8List salt = bytes.sublist(0, SALT_LENGTH);
    Uint8List iv = bytes.sublist(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
    String ciphertext = base64.encode(
        bytes.sublist(SALT_LENGTH + IV_LENGTH + TAG_LENGTH, bytes.length));
    Uint8List key = _pbkdf2(password, salt);
    return AesGcm(key).decrypt(ciphertext, iv);
  }

  static Uint8List _combineInputs(
      Uint8List salt, Uint8List iv, Uint8List tag, String ciphertext) {
    return Uint8List.fromList(salt + iv + tag + base64.decode(ciphertext));
  }

  /// Password Based Key Deriviation function
  static Uint8List _pbkdf2(String password, Uint8List salt) {
    PBKDF2 generator = new PBKDF2(sha512);
    return generator.generateKey(
        password, salt, KEY_ITERATIONS_COUNT, KEY_LENGTH);
  }

  /// Generates salt of given length and returns Uint8List result
  static Uint8List _generateTag() => _randomBytes(TAG_LENGTH);

  /// Generates salt of given length and returns Uint8List result
  static Uint8List _generateSalt() => _randomBytes(SALT_LENGTH);

  /// Generates initialization vector of given length and returns Uint8List result
  static Uint8List _generateIV() => _randomBytes(IV_LENGTH);

  /// Generates a random byte sequence of given [length]
  static Uint8List _randomBytes(int length) {
    Uint8List buffer = new Uint8List(length);
    Random range = new Random.secure();
    for (int i = 0; i < length; i++) {
      buffer[i] = range.nextInt(256);
    }
    return buffer;
  }
}
